### ISSUE TITLE
clang 14.0.6 on windows misdiagnoses valid code that clang 14.0.6 on linux compiles correctly

### ISSUE THREAD

**Issue by RoyAwesome** [llvm/llvm-project#56989]
So, the following code gives me a very incorrect error on windows, however clang-14 on linux works.  msvc compiles this code correctly

This code is cut down to just what is needed for the issue.  

```cpp
#include <tuple>
#include <type_traits>
#include <coroutine>

template<typename AwaitableContainer>
class when_all_ready_awaitable;


template<typename... Tasks>
class when_all_ready_awaitable<std::tuple<Tasks...>>
{          
public:

    explicit when_all_ready_awaitable(Tasks&&... tasks)
        : m_counter(sizeof...(Tasks))
        , m_tasks(std::move(tasks)...)
    { }

    auto operator co_await() && 
    {
        struct awaiter
        {
            awaiter(when_all_ready_awaitable& in_awaitable) 
                : m_awaitable(in_awaitable)                   
            { }

            awaiter() = delete;

            [[nodiscard]] bool await_ready() const noexcept
            {
                return true;
            }
            bool await_suspend(std::coroutine_handle<> in_coro) noexcept
            {
                return false;
            }

            std::tuple<Tasks...>&& await_resume() noexcept
            {
                return std::move(m_awaitable.m_tasks);
            }
        private:
            when_all_ready_awaitable& m_awaitable;
        };

        return awaiter { *this };
    }
private:
    int m_counter;
    std::tuple<Tasks...> m_tasks;
};

int main()
{
    when_all_ready_awaitable<std::tuple<int, int>> foo(10, 10);
    static_cast<when_all_ready_awaitable<std::tuple<int, int>>&&>(foo).operator co_await();

    return 0;
}
```

Compiled on clang 14.0.6 on Windows: 

```
clang++.exe -std=c++20 .\test.cpp
.\test.cpp:24:13: error: constructor for 'awaiter' must explicitly initialize the reference member 'm_awaitable'
            awaiter(when_all_ready_awaitable& in_awaitable)
            ^
.\test.cpp:22:16: note: in instantiation of member function 'when_all_ready_awaitable<std::tuple<int, int>>::operator
      co_await()::awaiter::awaiter' requested here
        struct awaiter
               ^
.\test.cpp:44:39: note: declared here
            when_all_ready_awaitable& m_awaitable;
                                      ^
.\test.cpp:41:34: error: 'when_all_ready_awaitable<std::tuple<int, int>>::operator
      co_await()::awaiter::await_resume()::awaiter::m_awaitable' is not a member of class 'awaiter'
                return std::move(m_awaitable.m_tasks);
                                 ^
.\test.cpp:22:16: note: in instantiation of member function 'when_all_ready_awaitable<std::tuple<int, int>>::operator
      co_await()::awaiter::await_resume' requested here
        struct awaiter
               ^
```

Where on Linux and msvc, this compiles correctly.  

Obviously, I explicitly initialize the m_awaitable variable in there.  This seems to be a bug for Clang on Windows.

**Comment by RoyAwesome** [comment]
I should note, moving the struct outside of the function causes the code to compile correctly on windows clang.

**Comment by ChuanqiXu9** [comment]
@RoyAwesome would you like to give it a test again? Since I don't have a windows develop environment.

**Comment by RoyAwesome** [comment]
Any particular version you want me to test against? I can test this on clang 16 later this evening

**Comment by ChuanqiXu9** [comment]
> Any particular version you want me to test against? I can test this on clang 16 later this evening

If it is possible, it will be best to test against trunk.
